homeassistant/components/datetime/__init__.py

```diff
@@ -0,0 +1,170 @@
+"""Component to allow setting date/time as platforms."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import datetime, timedelta, timezone
+import functools
+import logging
+from typing import final
+
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_DATE, ATTR_TIME
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    ENTITY_SERVICE_FIELDS,
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import Entity, EntityDescription
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.util import dt as dt_util
+
+from .const import ATTR_DATETIME, ATTR_TIME_ZONE, DOMAIN, SERVICE_SET_VALUE
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+_LOGGER = logging.getLogger(__name__)
+
+__all__ = ["DOMAIN", "DateTimeEntity", "DateTimeEntityDescription"]
+
+
+def _split_date_time(config):
+    """Split date/time components."""
+    datetime_ = config.pop(ATTR_DATETIME, None)
+    config[ATTR_DATE] = datetime_.date()
+    config[ATTR_TIME] = datetime_.time()
+    return config
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Date/Time entities."""
+    component = hass.data[DOMAIN] = EntityComponent[DateTimeEntity](
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_SET_VALUE,
+        vol.All(
+            vol.Any(
+                vol.All(
+                    vol.Schema(
+                        {
+                            vol.Optional(ATTR_DATE): cv.date,
+                            vol.Optional(ATTR_TIME): cv.time,
+                            vol.Optional(ATTR_TIME_ZONE): cv.time_zone,
+                            **ENTITY_SERVICE_FIELDS,
+                        }
+                    ),
+                    cv.has_at_least_one_key(ATTR_DATE, ATTR_TIME),
+                ),
+                vol.All(
+                    vol.Schema(
+                        {
+                            vol.Required(ATTR_DATETIME): cv.datetime,
```

Comment:
`cv.datetime` actually also supports timezones (the offsets format), for example: `2020-12-13 10:00:00+01:00` would be valid to use in this service call.

However, due to splitting date/time that gets lost and in `_async_set_value` it would end up setting the local system timezone instead at that point.


I think I would recommend turning it around and making this the main input method for the service call (and thus not split in this case, but combine in the other case).

Knowing that, one could argue the need for a separate `time_zone` parameter in this service call as well maybe?

Reply:
good point, so I will remove `time_zone` as a parameter. The split logic allows the user to specify just a `date` or just a `time` and to use the existing `time` or `date` value respectively as part of the update. Do we want to preserve this functionality? I don't think it's necessary, I just pulled the behavior from `input_datetime`. If you'd like to keep it, let me know, we just need to update the split logic to also grab the timezone

Reply:
ah so one issue is that the date/time selector doesn't include a timezone component. So if a user is using the UI, the only way to input timezone would be to have a separate parameter. Thoughts?

Reply:
> So if a user is using the UI to call the service, the only way to input timezone would be to have a separate parameter. Thoughts?

I think from the UI perspective, it is fine to assume local time.

Reply:
I think this isn't fully resolved at this point.

I think the UI perspective is fine, but we still not handling a given timezone. A service call providing the following: `2020-12-13 10:00:00+01:00`, will ignore the given timezone and overwrite it with the one configured in the system.

I wonder if we make this simpler by removing the separate date/time/timezone service version and just raw accept the single datetime, in case the timezone is missing from that object, replace it with the system one (otherwise skip it).

Reply:
why do you say that? we store the offset at `ATTR_OFFSET` so that we can reapply it in the service handler

Reply:
My oversight! 🙏